### PR TITLE
crud operations on routes

### PR DESCRIPTION
**Pull request by kwizeradev**
## CRUD Operations on routes

This PL is CRUD Operations on Routes using sequelize TYPEORM. 

- As an operator, I want to create routes so that buses can travel in pre-defined routes.
- As an operator, I want to create routes so that users can view in pre-defined routes.

# Start Up
Please clone the PR and start up by running the commands as shown below:

```console
git clone https://github.com/atlp-rwanda/Dominators-Phantom-BE.git -b ft-api-crud-routes-1103
cd Dominators-Phantom-BE

npm run dev
```

## Routes CRUD Operations

[Run in Postman](https://www.getpostman.com/collections/581d8c909c3901bc9f44)

| HTTP Verb  |                                               `REQUEST`                                               |                                              `Params`                                              |
| ---------- | :--------------------------------------------------------------------------------------------------------: | :------------------------------------------------------------------------------------------------------------------: |
| **POST**   |                               localhost:3000/routes/                               |                                 origin,destination,code,distance                                |
| **GET**    |localhost:3000/routes/| To Paginate : localhost:3000/routes?page=1&size=2|
| **PUT**    |                                                  localhost:3000/routes/routeSlug                                                  |                                                         :Optional:                                                          |
| **DELETE**  |                                                localhost:3000/routes/routeSlug                                                    |                    ⛔                                                                           |
| **DELETE ALL** |                                                   localhost:3000/routes/all/delete                                                     |                                          ⛔                                        |

# What to expect

- [x] The operator should be able to create, delete, and update a route.
- [x] The users should be able to view a route's details
- [x] Appropriate error messages are displayed on unsuccessful attempts

| HTTP Verb   |                           `Success`                           |                `Requirements`                |                              `Bad request`                               |
| ----------- | :-------------------------------------------------------------: | :---------------------------------------------------: | :--------------------------------------------------------------------------------------: |
| **POST**    |  Status:success, code:200 and route object                                |                          Route should be unique                         |                                           Status:fail, code:400, record:Message already exist                                           |
| **GET**     |                               List of all routes                               |                          Make sure you have some routes in database                          | Can not GET from localhost/routes|
| **PUT**     |                              "message": "Route was updated successfully."                          |                         Should have valid route Slug                         |              Cannot update Route with id=specifiedId               |

### CODE REVIEW

**Review by Kalimbak — APPROVED** [review]

**Review by Angelus123 — APPROVED** [review]

File: src/controllers/RouteController.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    //                 { origin: req.body.origin },
+    //                 { destination: req.body.destination }
+    //             ]
+    //         }
+    //     });
+    //     if (foundItem) {
+    //         return responseHandler(res, 400, "Conflicts occured with existing route.")
+    //     }
+    // }
+
```
Comment (furebo) [inline src/controllers/RouteController.js]: Remove the commented codes

File: src/controllers/RouteController.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            responseHandler(res, 500, err.message || "Some error occurred while creating the route.")
+        });
+}
+
+const findAll = async (req, res) => {
+
+    await busRoutes.findAll({})
+        .then(data => {
+            res.send(data);
+        })
```
Comment (furebo) [inline src/controllers/RouteController.js]: Add pagination

**Review by furebo — CHANGES_REQUESTED** [review]
Please remember also to translate the messages

File: src/routes/busRoutes.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { Router } from 'express';
+import { addRoute, findAll, findOne, updateRoute, removeRoute, deleteAll } from '../controllers/RouteController.js';
+const router = Router();
```
Comment (tjeaneric) [inline src/routes/busRoutes.js]: Can we move these routes in a folder called api?

**Review by tjeaneric — APPROVED** [review]

File: src/controllers/RouteController.js (hunk 1 above)
Comment (kwizeradev) [reply src/controllers/RouteController.js]: Done

File: src/controllers/RouteController.js (hunk 2 above)
Comment (kwizeradev) [reply src/controllers/RouteController.js]: Done

File: src/routes/busRoutes.js (hunk 3 above)
Comment (kwizeradev) [reply src/routes/busRoutes.js]: Done

File: src/app.js [hunk 4]
```diff
@@ -25,6 +49,7 @@ app.get('/api/v1', (req, res) => {
   console.log(req.t())
 
 });
+app.use('', appRoutes);
```
Comment (tjeaneric) [inline src/app.js]: correct routing

File: src/routes/index.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { Router } from 'express';
+import { busRoutes } from './api/busRoutes';
+
+const router = Router();
+
+
+router.use('/api/v1/routes', busRoutes);
```
Comment (tjeaneric) [inline src/routes/index.js]: correct your routes

File: src/server.js [hunk 6]
```diff
@@ -9,4 +15,4 @@ const server = app.listen(port, () =>
   console.log(`App listening on ${port}!....`)
```
Comment (tjeaneric) [inline src/server.js:15]: comfort??

**Review by tjeaneric — CHANGES_REQUESTED** [review]

**Review by furebo — CHANGES_REQUESTED** [review]
Rebase to resolve the conflicts and add authentication to the endpoints

File: locales/en/translation.json [hunk 7]
```diff
@@ -13,4 +23,8 @@
   "user_update": "User with ${id} was updated successfully",
   "uesr_exists": "User with id of ${id} does not exist",
   "user_deleted": "User with id of ${id} successfully deleted"
+<<<<<<< HEAD
```
Comment (tjeaneric) [inline locales/en/translation.json]: Resolve conflicts here.

File: package.json [hunk 8]
```diff
@@ -7,15 +7,12 @@
     "test": "better-npm-run test",
     "dev": "better-npm-run dev",
     "start": "npm run build && better-npm-run prod",
-    "migrate": "sequelize-cli db:migrate",
```
Comment (tjeaneric) [inline package.json]: Add these commands

File: package.json [hunk 9]
```diff
@@ -27,7 +24,7 @@
       }
     },
     "test": {
-      "command": "nyc mocha --recursive src/**/*.test.js --require @babel/register src/**/*_test.js __test__/",
```
Comment (tjeaneric) [inline package.json:30]: Why did you change the command for testing?

File: package.json [hunk 10]
```diff
@@ -7,15 +7,12 @@
...
     "start": "npm run build && better-npm-run prod",
-    "migrate": "sequelize-cli db:migrate",
-    "undo-migrate": "sequelize-cli db:migrate:undo",
     "build": "babel src -d dist"
   },
   "betterScripts": {
     "dev": {
       "command": "nodemon --exec babel-node ./src/server",
       "env": {
-        "NODE_ENV": "development",
```
Comment (tjeaneric) [inline package.json:18]: Add this NODE_ENV

File: src/app.js [hunk 11]
```diff
@@ -6,14 +6,25 @@ import Backend from 'i18next-fs-backend';
 import middleware from 'i18next-http-middleware';
 import globalErrorHandler from './controllers/errorController';
 import swaggerDocument from './documentation/index';
-import routes from './routes/index';
```
Comment (tjeaneric) [inline src/app.js]: Use this route

File: src/app.js [hunk 12]
```diff
@@ -25,14 +36,18 @@ i18next
...
 app.get('/api/v1', (req, res) => {
   res.status(200).json({
-    message: req.t('welcome_message'),
+    message: req.t('welcome_message')
   });
+  console.log(req.t())
+
 });
 
-app.use('/api/v1/', routes);
```
Comment (tjeaneric) [inline src/app.js]: You deleted others' codes

File: src/app.js [hunk 13]
```diff
@@ -45,7 +60,5 @@ app.use(
   })
 );
 
-//ERROR HANDLING MIDDLEWARE
-app.use(globalErrorHandler);
```
Comment (tjeaneric) [inline src/app.js:49]: You deleted others' codes

File: src/middlewares/auth.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import jwt from 'jsonwebtoken';
+import prop from '../config/config';
+const currentConfig = prop[process.env.NODE_ENV];
+
+const { JWT_SECRET } = currentConfig;
+
+
```
Comment (tjeaneric) [inline src/middlewares/auth.js]: this middleware is not needed, You can use protect in authController

File: src/routes/api/busRoutes.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { Router } from 'express';
+import { addRoute, findAll, findOne, updateRoute, removeRoute, deleteAll } from '../../controllers/RouteController.js';
+import { authUser } from '../../middlewares/auth';
+const router = Router();
+
+router.get('/', authUser, findAll);
```
Comment (tjeaneric) [inline src/routes/api/busRoutes.js]: use Protect in authController

File: src/routes/index.js [hunk 16]
```diff
@@ -1,8 +1,11 @@
-import express from 'express';
+import { Router } from 'express';
+import busRoutes from './api/busRoutes';
 import userRoutes from './api/userRoutes';
 
-const routes = express.Router();
```
Comment (tjeaneric) [inline src/routes/index.js:4]: You changed others' work

**Review by tjeaneric — CHANGES_REQUESTED** [review]

File: locales/en/translation.json (hunk 7 above)
Comment (Angelus123) [reply locales/en/translation.json]: And also remember to do reduce number of commit by (Squash)

**Review by Angelus123 — CHANGES_REQUESTED** [review]

**Review by tjeaneric — APPROVED** [review]

**Review by furebo — APPROVED** [review]
